core/src/main/java/org/apache/accumulo/core/file/rfile/GenerateSplits.java

```diff
@@ -0,0 +1,301 @@
+
+    Configuration hadoopConf = new Configuration();
+    SiteConfiguration siteConf = opts.getSiteConfiguration();
+    boolean encode = opts.base64encode;
+
+    TreeSet<String> splits;
+    TreeSet<String> desiredSplits = new TreeSet<>();
+
+    if (opts.numSplits > 0 && opts.splitSize > 0) {
+      throw new IllegalArgumentException("Requested number of splits and split size.");
+    }
+    if (opts.numSplits == 0 && opts.splitSize == 0) {
+      throw new IllegalArgumentException("Required number of splits or split size.");
+    }
+    long numSplits = opts.numSplits;
+    long splitSize = opts.splitSize;
+
+    FileSystem fs = FileSystem.get(hadoopConf);
+    FileSystem localFs = FileSystem.getLocal(hadoopConf);
+    List<Path> filePaths = new ArrayList<>();
+    for (String file : opts.files) {
+      Path path = new Path(file);
+      if (file.contains(":")) {
+        fs = path.getFileSystem(hadoopConf);
+      } else {
+        log.warn("Attempting to find file across filesystems. Consider providing URI "
+            + "instead of path");
+        if (!fs.exists(path))
+          fs = localFs; // fall back to local
+      }
+      // get all the files in the directory
+      if (fs.getFileStatus(path).isDirectory()) {
+        // can only explode one directory
+        if (opts.files.size() > 1)
+          throw new IllegalArgumentException("Only one directory can be specified");
+        var iter = fs.listFiles(path, true);
+        while (iter.hasNext()) {
+          filePaths.add(iter.next().getPath());
+        }
+      } else {
+        filePaths.add(path);
+      }
+    }
+
+    // if no size specified look at indexed keys first
+    if (opts.splitSize == 0) {
+      splits = getIndexKeys(siteConf, hadoopConf, fs, filePaths, encode);
+      // if there weren't enough splits indexed, try again with size = 0
+      if (splits.size() < numSplits) {
+        log.info("Only found {} indexed keys but need {}. Doing a full scan on files {}",
+            splits.size(), numSplits, filePaths);
+        splits = getSplitsBySize(siteConf, hadoopConf, filePaths, fs, 0, encode);
+      }
+    } else {
+      splits = getSplitsBySize(siteConf, hadoopConf, filePaths, fs, splitSize, encode);
+    }
+
+    int numFound = splits.size();
+    // its possible we found too many indexed so take every (numFound / numSplits) split
+    if (opts.splitSize == 0 && numFound > numSplits) {
+      var iter = splits.iterator();
+      double targetFactor = (double) numFound / numSplits;
+      log.debug("Found {} splits but requested {} picking 1 every {}", numFound, opts.numSplits,
+          targetFactor);
+
+      for (int i = 0; i < numFound; i++) {
+        double currFactor = (double) (i + 1) / desiredSplits.size();
+        String next = iter.next();
+        // unsure if this should be currFactor >= targetFactor
+        if (currFactor > targetFactor && desiredSplits.size() < numSplits) {
+          desiredSplits.add(next);
+        }
+      }
```

Comment:
The following fixes some of the issues I mentioned.  Its modeled after the code in TabletOperations.listSplits w/o the suspicious while loop.

```suggestion
      var iter = splits.iterator();
      // This is how much each of the found rows will advances twoards a desired split point.  Add one to numSplits because if we request 9 splits, there will 10 tablets and we the 9 splits evenly spaces between the 10.
      double increment = (numSplits+1.0)/numFound;
      log.debug("Found {} splits but requested {} picking 1 every {}", numFound, opts.numSplits,
          targetFactor);

       // Tracks how far along we are twoards the next split.
       double total = 0;

      for (int i = 0; i < numFound; i++) {
        total += increment;
        String next = iter.next();
        // unsure if this should be currFactor >= targetFactor
        if (total> 1 && desiredSplits.size() < numSplits) {
          desiredSplits.add(next);
          total -= 1;
        }
      }
```

I created the following little program locally to play with the code and it seems good.

```java
public class Test {
  public static void main(String[] args) {
    printSplits(100, 9);
    System.out.println();
    printSplits(15, 10);
    System.out.println();
    printSplits(10, 10);
  }

  private static void printSplits(int numFound, int numSplits) {
    // This is how much each of the found rows will advances twoards a desired split point.  Add one to numSplits because if we request 9 splits, there will 10 tablets and we the 9 splits evenly spaces between the 10.
    double increment = (numSplits+1.0)/numFound;

    System.out.println("increment "+increment);
    
    int added = 0;

    // Tracks how far along we are twoards the next split.
    double total = 0;
    
    for (int i = 0; i < numFound; i++) {
     total += increment;
    
      // unsure if this should be currFactor >= targetFactor
      if (total > 1 && added < numSplits) {
        System.out.println(i+" "+added +" "+total);
        added++;
        total -= 1;
      }
    }
  }
}
```

Reply:
> total -= 1;

This doesn't seem right to me. Should this be `total -= increment` or `total = 0`? This seems to add too many splits in a row.

Reply:
This fails the test case with 6 splits and `--num 4`. It will only return 3 splits, r2, r4, r6.

Reply:
Ignore that last message I was testing with a change I made to set total = 0.

Reply:
> Should this be total -= increment or total = 0? This seems to add too many splits in a row.

@milleruntime  the goal is if the total is 1.66 then the .66 will not be lost and can contribute towards future splits.  May be good to add a comment to the code about this.

Reply:
OK I understand it now. I added a comment, renamed some variables (I thought `progressToNextSplit` was more clear than `total`) and incorporated your test program into the unit test in c3d7205. Let me know if you can think of any other edge cases and I can add them to the test.